filters/apimonitoring/filter.go

```diff
@@ -0,0 +1,158 @@
+)
+
+const (
+	stateBagKeyPrefix = "filter.apimonitoring."
+	stateBagKeyState  = stateBagKeyPrefix + "state"
+)
+
+var (
+	unknownPath = &pathInfo{
+		ApplicationId: unknownElementPlaceholder,
+		ApiId:         unknownElementPlaceholder,
+		PathTemplate:  unknownElementPlaceholder,
+	}
+)
+
+type apiMonitoringFilter struct {
+	paths []*pathInfo
+}
+
+type pathInfo struct {
+	ApplicationId string
+	ApiId         string
+	PathTemplate  string
+	Matcher       *regexp.Regexp
+}
+
+// apiMonitoringFilterContext contains information about the metrics tracking
+// for one HTTP exchange (one routing). It serves to pass information from
+// the `Request` call to the `Response` call (stored in the context's `StateBag`).
+type apiMonitoringFilterContext struct {
+	// DimensionPrefix is the prefix to all metrics tracked during this exchange (generated only once)
+	DimensionsPrefix string
+	// Begin is the earliest point in time where the request is observed
+	Begin time.Time
+	// OriginalRequestSize is the initial requests' size, before it is modified by other filters.
+	OriginalRequestSize int64
+}
+
+func (f *apiMonitoringFilter) String() string {
+	return toJsonStringOrError(mapApiMonitoringFilter(f))
+}
+
+func (f *apiMonitoringFilter) Request(c filters.FilterContext) {
+	log := log.WithField("op", "request")
+
+	// Identify the dimensions "prefix" of the metrics.
+	dimensionsPrefix := f.getDimensionPrefix(c, log)
+
+	// Gathering information from the initial request for further metrics calculation
+	begin := time.Now()
+	originalRequestSize := c.Request().ContentLength
+
+	// Store that information in the FilterContext's state.
+	mfc := &apiMonitoringFilterContext{
+		DimensionsPrefix:    dimensionsPrefix,
+		Begin:               begin,
+		OriginalRequestSize: originalRequestSize,
+	}
+	c.StateBag()[stateBagKeyState] = mfc
+}
+
+func (f *apiMonitoringFilter) Response(c filters.FilterContext) {
+	log := log.WithField("op", "response")
+	log.Debugf("RESPONSE CONTEXT: %s", c)
+
+	mfc, ok := c.StateBag()[stateBagKeyState].(*apiMonitoringFilterContext)
+	if !ok {
+		log.Debugf("Call not tracked")
+		return
+	}
+
+	metrics := c.Metrics()
+	response := c.Response()
+
+	f.writeMetricCount(metrics, mfc)
+	f.writeMetricResponseStatusClassCount(metrics, mfc, response)
+	f.writeMetricLatency(metrics, mfc)
+}
+
+// getDimensionPrefix generates the dimension part of the metrics key (before the name
+// of the metric itself).
+func (f *apiMonitoringFilter) getDimensionPrefix(c filters.FilterContext, log *logrus.Entry) string {
+	req := c.Request()
+	var path *pathInfo = nil
+	for _, p := range f.paths {
+		if p.Matcher.MatchString(req.URL.Path) {
+			path = p
+			break
+		}
+	}
+	if path == nil {
+		log.Debugf("Matching no path template. Tracking as unknown.")
+		path = unknownPath
+	}
+	dimensions := []string{
+		path.ApplicationId,
+		path.ApiId,
+		req.Method,
+		path.PathTemplate,
+		"",
+	}
+	prefix := strings.Join(dimensions, metricsSeparator)
```

Comment:
@tkrop 

If you are referring to the `ApplicationId` and `ApiId` part at the beginning, it could indeed be pre-merged together. But that would not solve anything GB-pressure wise. Look at the implementation of `strings.Join`. It allocates just ONCE the proper size for the total string, which we still need to render dynamically since the `Method` comes before the `Path`.

This has to be dynamically _compiled_ somehow.

The only avoidable thing would be the array creation (`dimensions`) and would only make sense if we would change the order of metrics parts to start with the fix part `AppId.ApiId.Path.` (which can be rendered at configuration time) and then just add the _HTTP verb_ to it.

I honestly do not think it would be that much of an improvement.

But open to discussion.

Reply:
Solved in 8a62d0d80d0e4e154a9801a73e3234c99f1bcb0e